File: app/src/main/java/com/absinthe/kage/media/video/VideoPlayer.java
package com.absinthe.kage.media.video;

import android.content.Context;
import android.graphics.Bitmap;
import android.media.session.PlaybackState;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import android.widget.TextView;
import android.widget.VideoView;

import androidx.annotation.AttrRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.absinthe.kage.R;
import com.absinthe.kage.databinding.ViewVideoPlayerBinding;
import com.absinthe.kage.media.LocalMedia;
import com.absinthe.kage.media.Playback;

import java.util.Formatter;
import java.util.Locale;

public class VideoPlayer extends FrameLayout implements Playback.Callback {

    public static final int TYPE_LOCAL = 1;
    public static final int TYPE_REMOTE = 2;

    private static final String TAG = VideoPlayer.class.getSimpleName();

    private Context mContext;
    private ViewVideoPlayerBinding mBinding;
    private StringBuilder mFormatBuilder;
    private Formatter mFormatter;
    private ImageView mIVCover;
    private ImageView mIVPause;
    private Playback mPlayback;
    private View mRoot;
    private SeekBar mSeekBar;

    private final OnSeekBarChangeListener mOnSeekBarChangeListener = new OnSeekBarChangeListener() {
        public void onStartTrackingTouch(SeekBar seekBar) {
            mDragging = true;
            removeCallbacks(mShowProgress);
        }

        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            if (fromUser && mPlayback != null) {
                long newPosition = (((long) progress) * ((long) mPlayback.getDuration())) / 1000;
                mPlayback.seekTo((int) newPosition);
                if (mTVCurrentTime != null) {
                    mTVCurrentTime.setText(stringForTime((int) newPosition));
                }
            }
        }

        public void onStopTrackingTouch(SeekBar seekBar) {
            mDragging = false;
            setProgress();
            updatePausePlay();
            post(mShowProgress);
        }
    };
    private final OnClickListener mPauseListener = v -> doPauseResume();

    private boolean isLoaded;
    private boolean mDragging;
    private int mBeforePosition = 0;
    private int mPlayState = 0;

    private final Runnable mShowProgress = new Runnable() {
        public void run() {
            Log.d(TAG, "mShowProgress run");
            int pos = setProgress();
            if (!mDragging && isPlaying()) {
                postDelayed(mShowProgress, 1000 - (pos % 1000));
            }
        }
    };

    private LocalMedia mLocalMedia;
    private TextView mTVCurrentTime;
    private TextView mTVDurationTime;
    private VideoPlayCallback mVideoPlayCallback;
    private VideoView mVideoView;

    public interface VideoPlayCallback {
        void changeState(int i);
    }

    public VideoPlayer(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        mContext = context;
        init();
    }

    public VideoPlayer(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mContext = context;
        init();
    }

    private void init() {
        LayoutParams frameParams = new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
        removeAllViews();
        addView(makeControllerView(), frameParams);
    }

    protected View makeControllerView() {
        LayoutInflater mInflater = (LayoutInflater) mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        mBinding = ViewVideoPlayerBinding.inflate(mInflater);
        mRoot = mBinding.getRoot();
        initView();
        return mRoot;
    }

    private void initView() {
        mVideoView = mBinding.videoView;
        mIVCover = mBinding.ivCover;
        mSeekBar = mBinding.layoutSeekbar.seekbar.seekBar;
        mTVDurationTime = mBinding.layoutSeekbar.seekbar.tvDuration;
        mTVCurrentTime = mBinding.layoutSeekbar.seekbar.tvCurrentTime;
        mIVPause = mBinding.layoutSeekbar.ivPlay;
        mFormatBuilder = new StringBuilder();
        mFormatter = new Formatter(mFormatBuilder, Locale.getDefault());
        mIVPause.setOnClickListener(mPauseListener);
        mSeekBar.setOnSeekBarChangeListener(mOnSeekBarChangeListener);
        mVideoView.setOnErrorListener((mp, what, extra) -> what == Integer.MIN_VALUE || what == -38);
    }

    public void playMedia(LocalMedia localMedia) {
        removeCallbacks(mShowProgress);
        mBeforePosition = 0;
        mLocalMedia = localMedia;
        if (mPlayback != null) {
            mPlayback.playMedia(localMedia);
        }
    }

    public void changePlayer(int type) {
        if (mPlayback != null) {
            mBeforePosition = mPlayback.getCurrentPosition();
            mPlayback.stop(true);
            mPlayback.setCallback(null);
            mPlayback = null;
        }
        if (type == TYPE_LOCAL) {
            mIVCover.setVisibility(GONE);
            mVideoView.setVisibility(VISIBLE);
            mPlayback = new LocalVideoPlayback(mVideoView);
            mPlayback.setCallback(this);
        } else if (type == TYPE_REMOTE) {
            mIVCover.setVisibility(VISIBLE);
            mVideoView.setVisibility(GONE);
            mPlayback = new RemoteVideoPlayback();
            mPlayback.setCallback(this);
        }
        if (mPlayback != null) {
            if (mLocalMedia != null) {
                mPlayback.playMedia(mLocalMedia);
            }
        }
    }

    public void release() {
        Log.d(TAG, "release");
        mPlayState = 0;
        Playback playback = mPlayback;
        if (playback != null) {
            playback.stop(false);
        }
    }

    public void setVideoPlayCallback(VideoPlayCallback callback) {
        mVideoPlayCallback = callback;
    }

    private void updatePausePlay() {
        if (mRoot != null && mIVPause != null) {
            if (isPlaying()) {
                Log.i(TAG, "updatePausePlay playing");
                post(mShowProgress);
                mIVPause.setImageResource(R.drawable.ic_pause);
            } else {
                Log.i(TAG, "updatePausePlay pause");
                removeCallbacks(mShowProgress);
                mIVPause.setImageResource(R.drawable.ic_play_arrow);
            }
        }
    }

    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        if (!isLoaded) {
            requestCoverImage();
        }
    }

    private void requestCoverImage() {
        ImageView imageView = mIVCover;
        if (imageView != null && imageView.getWidth() > 0 && mLocalMedia != null) {
            Log.d(TAG, "mIVCover start load");
            Bitmap b = VideoHelper.getVideoCoverImage(mLocalMedia.getFilePath());
            isLoaded = true;
            mIVCover.setImageBitmap(b);
        }
    }

    public void onCompletion() {
        Log.d(TAG, "onCompletion");
        SeekBar seekBar = mSeekBar;
        seekBar.setProgress(seekBar.getMax());
        updatePausePlay();
    }

    public void onPlaybackStateChanged(int state) {
        if (mPlayback != null && mPlayState == PlaybackState.STATE_BUFFERING
                && state == PlaybackState.STATE_PLAYING && mBeforePosition > 0) {
            Log.d(TAG, "Seek to: " + mBeforePosition);
            mPlayback.seekTo(mBeforePosition);
        }
        if (mPlayState != state) {
            updatePausePlay();
            VideoPlayCallback videoPlayCallback = mVideoPlayCallback;
            if (videoPlayCallback != null) {
                videoPlayCallback.changeState(state);
            }
            mPlayState = state;
        }
    }

    public void onError(String error) {
    }

    public void onMediaMetadataChanged(LocalMedia localMedia) {
        updatePausePlay();
        post(mShowProgress);
    }

    private int setProgress() {
        int position = mPlayback.getCurrentPosition();
        if (position == 0) {
            return 0;
        }

        int duration = mPlayback.getDuration();
        if (mSeekBar != null) {
            if (duration > 0) {
                mSeekBar.setProgress((int) ((((long) position) * 1000) / ((long) duration)));
            }
            mSeekBar.setSecondaryProgress(mPlayback.getBufferPosition() * 10);
        }
        TextView textView = mTVDurationTime;
        if (textView != null) {
            textView.setText(stringForTime(duration));
        }
        textView = mTVCurrentTime;
        if (textView != null) {
            textView.setText(stringForTime(position));
        }
        return position;
    }

    private void doPauseResume() {
        if (mPlayback != null) {
            if (isPlaying()) {
                mPlayback.pause();
            } else {
                mPlayback.play();
            }
            updatePausePlay();
            post(mShowProgress);
        }
    }

    private String stringForTime(int timeMs) {
        int totalSeconds = timeMs / 1000;
        int seconds = totalSeconds % 60;
        int minutes = (totalSeconds / 60) % 60;
        int hours = totalSeconds / 3600;
        mFormatBuilder.setLength(0);

        if (hours > 0) {
            return mFormatter.format("%d:%02d:%02d", hours, minutes, seconds).toString();
        }
        return mFormatter.format("%02d:%02d", minutes, seconds).toString();
    }

    private boolean isPlaying() {
        return mPlayback != null && mPlayback.getState() == PlaybackState.STATE_PLAYING;
    }
}

File: app/src/main/kotlin/com/absinthe/kage/media/video/VideoPlayer.kt
package com.absinthe.kage.media.video

import android.content.Context
import android.media.session.PlaybackState
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.annotation.AttrRes
import com.absinthe.kage.R
import com.absinthe.kage.databinding.ViewVideoPlayerBinding
import com.absinthe.kage.media.LocalMedia
import com.absinthe.kage.media.Playback
import com.absinthe.kage.media.TYPE_LOCAL
import com.absinthe.kage.media.TYPE_REMOTE
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.trackselection.AdaptiveTrackSelection
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import com.google.android.exoplayer2.ui.PlayerView
import com.google.android.exoplayer2.upstream.DefaultBandwidthMeter
import timber.log.Timber
import java.util.*

open class VideoPlayer : FrameLayout, Playback.Callback {

    private lateinit var mBinding: ViewVideoPlayerBinding
    private lateinit var mRoot: View
    private lateinit var seekBar: SeekBar
    private lateinit var playerView: PlayerView
    private lateinit var simpleExoPlayer: SimpleExoPlayer
    private lateinit var mPlayback: Playback

    private var mLocalMedia: LocalMedia? = null
    private var mVideoPlayCallback: VideoPlayCallback? = null

    private var mContext: Context
    private var mFormatBuilder: StringBuilder = StringBuilder()
    private var mFormatter: Formatter = Formatter(mFormatBuilder, Locale.getDefault())
    private var mDragging = false
    private var mBeforePosition = 0
    private var mPlayState = 0

    private val mShowProgress: Runnable = object : Runnable {
        override fun run() {
            val pos = setProgress()
            if (!mDragging && isPlaying) {
                postDelayed(this, 1000 - (pos % 1000).toLong())
            }
        }
    }

    private val mOnSeekBarChangeListener: OnSeekBarChangeListener = object : OnSeekBarChangeListener {

        override fun onStartTrackingTouch(seekBar: SeekBar) {
            mDragging = true
            removeCallbacks(mShowProgress)
        }

        override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
            if (fromUser) {
                val newPosition = progress.toLong() * mPlayback.duration.toLong() / 1000
                mPlayback.seekTo(newPosition.toInt())
                mBinding.layoutSeekBar.seekBar.tvCurrentTime.text = stringForTime(newPosition.toInt())
            }
        }

        override fun onStopTrackingTouch(seekBar: SeekBar) {
            mDragging = false
            setProgress()
            updatePausePlay()
            post(mShowProgress)
        }
    }

    private val isPlaying: Boolean
        get() = mPlayback.state == PlaybackState.STATE_PLAYING

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        mContext = context
        init()
    }

    constructor(context: Context, attrs: AttributeSet?, @AttrRes defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        mContext = context
        init()
    }

    private fun init() {
        val frameParams = LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        removeAllViews()
        addView(makeControllerView(), frameParams)
    }

    private fun makeControllerView(): View {
        val mInflater = mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        mBinding = ViewVideoPlayerBinding.inflate(mInflater)
        mRoot = mBinding.root
        initView()

        return mRoot
    }

    private fun initView() {
        seekBar = mBinding.layoutSeekBar.seekBar.seekBar
        playerView = mBinding.videoView

        mBinding.layoutSeekBar.ibPlay.setOnClickListener { doPauseResume() }
        seekBar.max = 1000
        seekBar.setOnSeekBarChangeListener(mOnSeekBarChangeListener)
        initExoPlayer()
    }

    private fun initExoPlayer() {
        val bandwidthMeter = DefaultBandwidthMeter.Builder(context).build()
        val videoTrackSelectionFactory = AdaptiveTrackSelection.Factory()
        val trackSelector = DefaultTrackSelector(context, videoTrackSelectionFactory)

        simpleExoPlayer = SimpleExoPlayer.Builder(context)
                .setBandwidthMeter(bandwidthMeter)
                .setTrackSelector(trackSelector)
                .build()
        playerView.player = simpleExoPlayer
        playerView.useController = false
        mPlayback = LocalVideoPlayback(simpleExoPlayer)
    }

    fun playMedia(localMedia: LocalMedia) {
        removeCallbacks(mShowProgress)
        mBeforePosition = 0
        mLocalMedia = localMedia
        mPlayback.playMedia(localMedia)
    }

    fun setPlayerType(type: Int) {
        mBeforePosition = mPlayback.currentPosition
        mPlayback.stop(true)

        if (type == TYPE_LOCAL) {
            mPlayback = LocalVideoPlayback(simpleExoPlayer)
            (mPlayback as LocalVideoPlayback).setCallback(this)
        } else if (type == TYPE_REMOTE) {
            mPlayback = RemoteVideoPlayback()
            (mPlayback as RemoteVideoPlayback).setCallback(this)
        }
        if (mLocalMedia != null) {
            mPlayback.playMedia(mLocalMedia!!)
        }
    }

    fun release() {
        Timber.d("release")
        mPlayState = 0
        mPlayback.stop(false)
        simpleExoPlayer.stop()
        simpleExoPlayer.release()
        LocalVideoPlayback.INSTANCE = null
    }

    fun setVideoPlayCallback(callback: VideoPlayCallback?) {
        mVideoPlayCallback = callback
    }

    fun setCastButtonClickListener(listener: OnClickListener) {
        mBinding.layoutSeekBar.btnCast.setOnClickListener(listener)
    }

    fun setCastButtonVisibility(flag: Int) {
        mBinding.layoutSeekBar.btnCast.visibility = flag
    }

    private fun updatePausePlay() {
        if (isPlaying) {
            post(mShowProgress)
            mBinding.layoutSeekBar.ibPlay.setImageResource(R.drawable.ic_pause)
        } else {
            removeCallbacks(mShowProgress)
            mBinding.layoutSeekBar.ibPlay.setImageResource(R.drawable.ic_play_arrow)
        }
    }

    override fun onCompletion() {
        Timber.d("onCompletion")
        seekBar.progress = seekBar.max
        updatePausePlay()
    }

    override fun onPlaybackStateChanged(state: Int) {
        if (mPlayState == PlaybackState.STATE_BUFFERING && state == PlaybackState.STATE_PLAYING && mBeforePosition > 0) {
            Timber.d("Seek to: $mBeforePosition")
            mPlayback.seekTo(mBeforePosition)
        }
        if (mPlayState != state) {
            updatePausePlay()
            mVideoPlayCallback?.changeState(state)
            mPlayState = state
        }
    }

    override fun onError(error: String) {}

    override fun onMediaMetadataChanged(localMedia: LocalMedia) {
        updatePausePlay()
        post(mShowProgress)
    }

    private fun setProgress(): Int {
        val position = mPlayback.currentPosition
        if (position == 0) {
            return 0
        }

        val duration = mPlayback.duration
        if (duration > 0) {
            seekBar.progress = (position.toLong() * 1000 / duration.toLong()).toInt()
        }
        seekBar.secondaryProgress = mPlayback.bufferPosition * 10
        mBinding.layoutSeekBar.seekBar.tvDuration.text = stringForTime(duration)
        mBinding.layoutSeekBar.seekBar.tvCurrentTime.text = stringForTime(position)
        return position
    }

    private fun doPauseResume() {
        if (isPlaying) {
            Timber.d("doPauseResume")
            mPlayback.pause()
        } else {
            mPlayback.play()
        }
        updatePausePlay()
        post(mShowProgress)
    }

    private fun stringForTime(timeMs: Int): String {
        val totalSeconds = timeMs / 1000
        val seconds = totalSeconds % 60
        val minutes = totalSeconds / 60 % 60
        val hours = totalSeconds / 3600
        mFormatBuilder.setLength(0)
        return if (hours > 0) {
            mFormatter.format("%d:%02d:%02d", hours, minutes, seconds).toString()
        } else mFormatter.format("%02d:%02d", minutes, seconds).toString()
    }

    interface VideoPlayCallback {
        fun changeState(state: Int)
    }
}